subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy

```diff
@@ -0,0 +1,429 @@
+                }
+                launcher2 = javaToolchains.launcherFor {
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata2.languageVersion.majorVersion})
+                }
+            }
+        """.stripIndent()
+
+        def installationPaths = [jdkMetadata1, jdkMetadata2].collect { it.javaHome.toAbsolutePath().toString() }.join(",")
+
+        when:
+        runWithInstallationPaths(installationPaths, task)
+        def events = eventsFor(taskPath)
+        def events1 = filterByJavaVersion(events, jdkMetadata1)
+        def events2 = filterByJavaVersion(events, jdkMetadata2)
+        then:
+        executedAndNotSkipped(taskPath)
+        events.size() > 0
+        assertToolchainUsages(events1, jdkMetadata1, "JavaLauncher")
+        assertToolchainUsages(events2, jdkMetadata2, "JavaLauncher")
+
+        // TODO: should it work with up-to-date tasks?
+//        when:
+//        runWithInstallationPaths(installationPaths, task)
+//        events = eventsFor(taskPath)
+//        events1 = filterByJavaVersion(events, jdkMetadata1)
+//        events2 = filterByJavaVersion(events, jdkMetadata2)
+//        then:
+//        skipped(taskPath)
+//        events.size() > 0
+//        assertToolchainUsages(events1, jdkMetadata1, "JavaLauncher")
+//        assertToolchainUsages(events2, jdkMetadata2, "JavaLauncher")
+    }
+
+    // TODO: test with two tasks using own toolchain in one build, and each event should be attributed to the right tasks
+    //  - two tasks could be in the same project
+    //  - two tasks could in different sub-projects (one using Java 8 and Java 11)
+
+    // TODO: javaCompile.options.forkOptions.javaHome = "blah" while using a toolchain
+
+    // TODO: test with Kotlin plugin
+
+    def "emit toolchain usages when task fails for 'compileJava' task"() {
+        def task = "compileJava"
+        def taskPath = ":$task"
+
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentJdk)
+
+        configureToolchainViaJavaPlugin(jdkMetadata)
+
+        file("src/main/java/Foo.java") << """
+            public class Foo extends Oops { }
+        """.stripIndent()
+
+        when:
+        /* TODO:
+            The error stream of the build-under-test has an unexpected error:
+            * What went wrong:
+                Execution failed for task ':compileJava'.
+                > Error while evaluating property 'javaCompiler' of task ':compileJava'
+                   > Failed to calculate the value of task ':compileJava' property 'javaCompiler'.
+                      > No compatible toolchains found for request filter: {languageVersion=8, vendor=any, implementation=vendor-specific} (auto-detect false, auto-download false)
+        */
+        runWithInstallationExpectingFailure(jdkMetadata, task)
+        def events = eventsFor(taskPath)
+        then:
+        fails(taskPath)
```

Comment:
The problem here is that `fails(taskPath)` runs the build again. Since you don't use `runWithInstallationExpectingFailure` it can't find the installation.  Something like this should do what you want:
```
        failureDescriptionStartsWith("Execution failed for task ':compileJava'.")
        failureHasCause("Compilation failed; see the compiler error output for details.")
```

I fixed this directly and pushed my changes.